crates/oxc_traverse/src/context/scoping.rs

```diff
@@ -193,6 +193,38 @@ impl TraverseScoping {
         new_scope_id
     }
 
+    /// Insert a scope between a parent and a child scope.
+    ///
+    /// For example, given the following scopes
+    /// ```ts
+    /// parentScope1: {
+    ///     childScope: { }
+    ///     childScope2: { }
+    /// }
+    /// ```
+    /// and calling this function with `parentScope1` and `childScope`,
+    /// the resulting scopes will be:
+    /// ```ts
+    /// parentScope1: {
+    ///     newScope: {   
+    ///         childScope: { }
+    ///     }
+    ///     childScope2: { }
+    /// }
+    /// ```
+    pub fn insert_scope_between(
+        &mut self,
+        parent_id: ScopeId,
+        child_id: ScopeId,
+        flags: ScopeFlags,
+    ) -> ScopeId {
+        let scope_id = self.create_child_scope(parent_id, flags);
+
+        self.scoping.remove_child_scopes(parent_id, &[child_id]);
+        self.scoping.set_scope_parent_id(child_id, Some(scope_id));
+        scope_id
+    }
```

Comment:
Please add a "shortcut" method to `TraverseCtx` too.

It would be preferable to add a new method `remove_child_scope` (singular) to `Scoping`. That method can be more efficient than `remove_child_scopes` because it only needs to search the `Vec` of children up until it finds the `ScopeId` it's looking for, whereas `remove_child_scopes` uses `retain` which carries on looking for more matches even after its found the first - which is pointless because there can be only 1 match.

Then:

```diff
- self.scoping.remove_child_scopes(parent_id, &[child_id]);
+ self.scoping.remove_child_scope(parent_id, child_id);
```

There should probably be a `debug_assert!` in this method checking that `parent_id` is indeed the parent of `child_id`.

Reply:
Also the call to `remove_child_scopes` can be gated with `if self.scoping.has_scope_child_ids() { ... }`.

Reply:
yep this makes sense to me, will fix, thanks